Add inbox_mentions setting

untitaker opened the pull request:
Since people keep mentioning cadey in the friendlysock thread, this
seems like a good time to add a setting.

This is a really basic subset of https://github.com/lobsters/lobsters/issues/872

I can add the other booleans as well, but thought i'd start with the most minimal needed change.

pushcx commented:
Thanks for taking this on, a partial solution seems worth implementing.

I think you did this about the same time I wrote 917f0a1c3c215e0ac36fef86a14b8c0429a70af1 so I see there's a conflict, and it's not using the pattern I tried to document there. We want to always create a Notification and then that class has a method to filter them out based on user settings (and events in the thread). The means that as the user changes their settings, those Notifications immediately disappear or appear as desired.

The name 'good faith' and 'bad faith' in the Notification class are a little misleading and reflects that they can also be hidden by events in the comment thread like flags. If you have ideas to better reflect the full functionality and would make it obvious how they're to be used, that would be a help here.

untitaker commented:
I think I addressed your review, but I wonder:

> The means that as the user changes their settings, those Notifications immediately disappear or appear as desired.

this doesn't seem fully desirable to me, if i turn on inbox_mentions after many years of having it off, my inbox is completely backlogged?

untitaker commented:
ehh this works but there is still a bug with the red notification count next to `Inbox`.

pushcx commented:
That bug is #1140.

untitaker commented:
So looking at your comments in #1720 i feel like I should just mark the notification instantly as read if not `should_display?`, and that'd fix a good chunk of issues like #1140?

pushcx commented on app/models/notification.rb:
```diff
@@ -9,20 +9,41 @@ class Notification < ApplicationRecord
+    comment = notifiable
+
+    # Check if this is a mention notification
+    if comment.plaintext_comment.match?(/\B[@~]#{Regexp.escape(user.username)}\b/)
```
I just wrote this exact same regexp in `extras/markdowner.rb`, let me lift it up to a constant you can reuse.

pushcx replied:
20d85ebf

pushcx commented:
Yeah, that's a clean way to handle it. (Sorry such a simple answer took so long, I was traveling.)

I added a constant for you to reuse in 20d85ebf. With those two small changes, this is ready to go.

Thanks for improving the name of those functions.

pushcx commented on app/models/notification.rb:
```diff
@@ -9,20 +9,41 @@ class Notification < ApplicationRecord
+    comment = notifiable
+
+    # Check if this is a mention notification
+    if comment.plaintext_comment.match?(/\B[@~]#{Regexp.escape(user.username)}\b/)
```
```suggestion
    if comment.plaintext_comment.match?(Markdowner::USERNAME_MENTION)
```

pushcx commented:
Thanks for coding this, @untitaker. I made the last couple small tweaks to get this merged before merge conflicts can really start accumulating or stalebot comes around to hassle.
